### PR TITLE
Get user by external id

### PR DESCRIPTION
**Pull request by johnnyaug**
1. User cache improvement: unify separate get user methods.
2. Addition to AuthService API: allow getting user by external ID.

### CODE REVIEW

**Comment by itaiad200** [comment]
Added @N-o-Z who's working in the same areas.

File: pkg/auth/service.go [hunk 1]
```diff
@@ -279,70 +280,63 @@ func (s *KVAuthService) DeleteUser(ctx context.Context, username string) error {
...
 				return nil, fmt.Errorf("failed to cast: %w", err)
 			}
-			if value.Email == email {
+			if predicate(value) {
 				return model.UserFromProto(value), nil
 			}
 		}
+		if itr.Err() != nil {
+			return nil, itr.Err()
+		}
```
Comment (guy-har) [inline pkg/auth/service.go:305]: Nice catch

**Review by eden-ohana — APPROVED** [review]

File: pkg/auth/db_service.go [hunk 2]
```diff
@@ -234,17 +243,31 @@ func (s *DBAuthService) GetUser(ctx context.Context, username string) (*model.Us
...
+		}, db.ReadOnly())
+		if err != nil {
+			return nil, err
+		}
+		return user.(*model.User), nil
+	})
+}
+
+func (s *DBAuthService) GetUserByExternalID(ctx context.Context, externalID string) (*model.User, error) {
+	return s.cache.GetUser(&userKey{externalID: externalID}, func() (*model.User, error) {
```
[the remark this answers is not in the document]
Comment (johnnyaug) [reply pkg/auth/db_service.go:254]: We're good here

File: pkg/auth/db_service.go [hunk 3]
```diff
@@ -234,17 +243,31 @@ func (s *DBAuthService) GetUser(ctx context.Context, username string) (*model.Us
...
 
 func (s *DBAuthService) GetUserByEmail(ctx context.Context, email string) (*model.User, error) {
-	user, err := s.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		return getDBUserByEmail(tx, email)
-	}, db.ReadOnly())
-	if err != nil {
-		return nil, err
-	}
-	return user.(*model.User), nil
+	return s.cache.GetUser(&userKey{email: email}, func() (*model.User, error) {
```
[the remark this answers is not in the document]
Comment (johnnyaug) [reply pkg/auth/db_service.go]: Fixed

**Review by guy-har — APPROVED** [review]

**Pull request merged**